controllers/interfaces.go

```diff
@@ -0,0 +1,333 @@
+}
+
+// fileExtractor provides functionality to get the contents of files from a container.
+//
+//nolint:unused
+//goland:noinspection GoUnusedType
+type fileExtractor interface {
+	// ExtractK8sResourcesFromContainer copies a file from stdout into map of strings.
+	ExtractK8sResourcesFromContainer(ctx context.Context, k8sExecPod exec.ExecPod) (map[string]string, error)
+}
+
+// applier provides ways to apply unstructured Kubernetes resources against the API.
+//
+//nolint:unused
+//goland:noinspection GoUnusedType
+type applier interface {
+	// ApplyWithOwner provides a testable method for applying generic, unstructured K8s resources to the API
+	ApplyWithOwner(doc apply.YamlDocument, namespace string, resource metav1.Object) error
+}
+
+//nolint:unused
+//goland:noinspection GoUnusedType
+type eventRecorder interface {
+	record.EventRecorder
+}
+
+// exposePortRemover is used to delete the exposure of the exposed services from the dogu.
+//
+//nolint:unused
+//goland:noinspection GoUnusedType
+type exposePortRemover interface {
+	// RemoveExposedPorts deletes the exposure of the exposed services from the dogu.
+	RemoveExposedPorts(ctx context.Context, doguResource *v2.Dogu, dogu *cesappcore.Dogu) error
+}
+
+type doguConfigRepository interface {
+	Get(ctx context.Context, name config.SimpleDoguName) (config.DoguConfig, error)
+	Create(ctx context.Context, doguConfig config.DoguConfig) (config.DoguConfig, error)
+	Update(ctx context.Context, doguConfig config.DoguConfig) (config.DoguConfig, error)
+	SaveOrMerge(ctx context.Context, doguConfig config.DoguConfig) (config.DoguConfig, error)
+	Delete(ctx context.Context, name config.SimpleDoguName) error
+	Watch(ctx context.Context, dName config.SimpleDoguName, filters ...config.WatchFilter) (<-chan repository.DoguConfigWatchResult, error)
+}
+
+// dependencyValidator checks if all necessary dependencies for an upgrade are installed.
+//
+//nolint:unused
+//goland:noinspection GoUnusedType
+type dependencyValidator interface {
+	// ValidateDependencies is used to check if dogu dependencies are installed.
+	ValidateDependencies(ctx context.Context, dogu *cesappcore.Dogu) error
+}
+
+// resourceUpserter includes functionality to generate and create all the necessary K8s resources for a given dogu.
+//
+//nolint:unused
+//goland:noinspection GoUnusedType
+type resourceUpserter interface {
+	resource.ResourceUpserter
+}
+
+// execPod provides methods for instantiating and removing an intermediate pod based on a Dogu container image.
+//
+//nolint:unused
+//goland:noinspection GoUnusedType
+type execPod interface {
+	exec.ExecPod
+}
+
+// execPodFactory provides functionality to create ExecPods.
+//
+//nolint:unused
+//goland:noinspection GoUnusedType
+type execPodFactory interface {
+	exec.ExecPodFactory
+}
+
+type podInterface interface {
+	v1.PodInterface
+}
+
+// premisesChecker includes functionality to check if the premises for an upgrade are met.
+type premisesChecker interface {
```

Comment:
🚫 **[checkstyle]** <unused> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>type `premisesChecker` is unused
<!-- __reviewdog__:ChBlZGU5MmQ4OWNhZTBkNzEwEgpjaGVja3N0eWxl -->